_posts/2018-10-15-12-2-Use-SkipToken-for-Server-Side-Paging.md

```diff
@@ -0,0 +1,63 @@
+### Proposed Solution
+WebAPI will now implement $skiptoken. When a collection of entity is requested which requires paging, we will assign the key value of the last sent entity to $skiptoken in the nextlink url. While processing a request with $skiptoken, we will add another condition (the key of the entity to be greater than the value specified to the skiptoken) to the predicate. 
+### Technical details
+After all the query options have been applied, we determine if the results need pagination. If the results need pagination, we will pass the key value of the last result to the method that generates the nextpage link.   
+
+#### Format of the nextlink
+The nextlink may contain $skiptoken if the result needs to be paginated. In WebAPI the $skiptoken value will be the key property and key value separated by a delimiter(:). For entities with composite or multi-part keys, each key property and value pair will be comma separated.
+```
+~/Products?$skiptoken=Id:27
+~/Books?$skiptoken=ISBN:978-2-121-87758-1,CopyNumber:11
+~/Products?$skiptoken=Id:25&$top=40
+```
+We will not use $skiptoken if the requested resource is not an entity type. Rather, normal skip will be used. 
+
+#### Generating the nextlink
+The next link generation method in ___GetNextPageHelper___ static class will take in the $skiptoken value along with other query parameters and generate the link by doing special handling for $skip, $skiptoken and $top. It will pass on the other query options as they were in the original request.
+##### 1. Handle $skip
+We will omit the $skip value if the service is configured to support $skiptoken and a collection of entity is being requested. This is because the first response would have applied the $skip query option to the results already. 
+##### 2. Handle $top
+We will reduce the value of $top query option by the page size if it is greater than the page size.   
+##### 3. Handle $skiptoken
+The value for the $skiptoken will be updated to new value passed in which is the key value for the last record sent. If the skiptoken value is not sent, we will revert to the previous logic and use $skip for paging instead.
+
+#### Parsing $skiptoken and generating the Linq expression
+New classes will be created for ___SkipTokenQueryOption___ and __SkipTokenQueryValidator__. ___SkipTokenQuery___ option will contain the  methods to create and apply the LINQ expression based on the $skiptoken value.
+#### Generating the $skiptoken
+The ___SkipTokenQueryOption___ class will be utilized by ___ODataQueryOption___ to pass the token value to the nextlink generator helper methods.
+In the process, ___IWebApiRequestMessage___ will be modified and GetNextPageLink method will now accept an optional parameter for the $skiptoken value.
```

Comment:
don't use the "optional parameter" otherwise, you can add new overload version.

Reply:
Done.